Implement Startup hooks support in Mono

lambdageek opened the pull request:
Fixes #47462 

**CoreCLR** This also makes some changes to CoreCLR to decouple EventPipe and startup hooks. Presently if startup hooks are disabled, `RuntimeEventSource.Initialize` is never called.  The PR makes the two features independent by moving runtime event source initialization out of the startup hook feature check.
TODO:
- [x] Make ios functional test
- [x] Make android functional test 
- [x] Make wasm functional test
- [x] Make an Issue to add mobile and wasm workload samples (once we have WBT for mobile) ==> https://github.com/dotnet/runtime/issues/80742
- [x] Fix error handling - the startup hooks design doc says the app should terminate if a hook has an unhandled exception. Right now mono cleans up the error and ignores it.

lambdageek committed:
Implement startup hooks support in Mono

am11 commented on src/mono/System.Private.CoreLib/src/ILLink/ILLink.Descriptors.xml:
```diff
@@ -643,5 +643,10 @@
                     <method name="GetInstanceFieldFieldStore" />
                 </type>
+
+                <!-- FIXME: only if feature="System.StartupHookProvider.IsSupported" featurevalue="true" -->
```
Something like:
```suggestion
  </assembly>

	<assembly fullname="System.Private.CoreLib" feature="System.StartupHookProvider.IsSupported" featurevalue="true" featuredefault="true">
```
?

lambdageek replied:
featuredefault is what I was missing. Thanks!

lambdageek committed:
Keep StartupHookProvider.ProcessStartupHooks under feature flag

lambdageek committed:
Don't catch/cleanup the exceptions from startup hooks.

lambdageek committed:
Add an ios simulator startup hook functional test

lambdageek committed:
Only initialize RuntimeEventSource if perftracing is supported

lambdageek committed:
Actually implement the ios startup hook test

vargaz commented:
The mono changes look ok to me.
Would this have any impact on linked app size and startup time ?

lambdageek commented:
> The mono changes look ok to me. Would this have any impact on linked app size and startup time ?

Linked size shouldn't be affected.  Unless you set `<StartupHookSupport>true</StartupHookSupport>` the .NET SDK will set `IsSupported` to false, and the managed code should drop out.

Startup time might be affected a little bit because of the `StartupHookProvider` lookup.

lambdageek committed:
Implement Android functional test

lambdageek committed:
Add WASM functional test

vitek-karas reviewed (approved):
The CoreLib changes look good.

vitek-karas commented:
Curious about the trimming story - in CoreCLR we simply don't support startup hooks with trimming. But this seems to make some effort to do that in mono. The problem is that startup hooks are effectively plugins, and so with trimmed framework the plugins may break.

lambdageek commented:
> Curious about the trimming story - in CoreCLR we simply don't support startup hooks with trimming. But this seems to make some effort to do that in mono. The problem is that startup hooks are effectively plugins, and so with trimmed framework the plugins may break.

The primary use case for mobile right now is to allow hot reload to inject an agent into the app. Injecting the hook is actually a pretty static operation - the agent assembly has to make it into the app bundle that gets shipped to the device - so there is a PackageReference, the SupportStartupHooks SDK property is toggled (which toggles the feature flag), etc. So everything should be visible to the trimmer if it even runs (in Debug configurations it won't be used).

The only other trimming that is happening is the initial trimming of System.Private.CoreLib when we create the runtime packs.  At that stage we say that runtime hook switch defaults to true and we add the process hooks function to the linker descriptor so it's kept around.

I think any sort of more dynamic startup hook injection is unlikely to work on mobile.  There's really no good way to get code on a device without it being part of the app bundle.  But running code before the user (or platform) code starts running is still useful.

lambdageek commented:
@lateralusX @dotnet/dotnet-diag

So believe it or not the failure I'm seeing is due to the call in `StartupHookProvider` to `RuntimeEventSource.Initialize()`.  Turns out mono already invokes that method from `ep_finish_init`.  So when we run it a second time, we instantiate a second `RuntimeEventSource` instance and then things go wrong.

Questions: 

1. As far as I can tell, coreclr only calls `RuntimeEventSource.Initialize` from `StartupHookProvider`.  Which means if `!StartupHookProvider.IsSupported`, the runtime event source is not initialized? Is that intentional? I could make mono do the same thing, but it seems weird to couple these two features. Or is there a second call site in coreclr somewhere?
2. Alternately, I could make `RuntimeEventSource.Initialize` safe to call multiple times by only creating a new instance if there is not one already (probably using a lock-object variant of `LazyInitializer.EnsureInitialized`)
3. Alternately, I could add `#if !MONO` to StartupHookProvider and leave both runtimes with the same behavior we have today

lambdageek committed:
Make RuntimeEventSource.Initialize safe to call multiple times

On Mono the method is called from ep_rt_mono_init_finish and also from
StartupHookProvider.ProcessStartupHooks.

CoreCLR only calls it from StartupHookProvider.  So if startup hooks
are disabled, the runtime event source is also disabled.

noahfalk commented:
> As far as I can tell, coreclr only calls RuntimeEventSource.Initialize from StartupHookProvider. Which means if !StartupHookProvider.IsSupported, the runtime event source is not initialized? Is that intentional?

No, that doesn't sound like a desired behavior to me. I don't see any reason that RuntimeEventSource should be conditional on having startup hooks enabled. I'm guessing what happened is that someone noticed it was important to initialize RuntimeEventSource prior to running startup hooks, and then separately someone else added an option to disable startup hooks and didn't recognize that the RuntimeEventSource.Initialize() should not be included in the scope of code that gets disabled. I'd propose we hoist the call to RuntimeEventSource.Initialize() outside the ProcessStartupHooks call.

lambdageek commented:
> I'd propose we hoist the call to RuntimeEventSource.Initialize() outside the ProcessStartupHooks call.

So change this to something more generic and call both the RuntimeEventSource initialization and the startup hook initialization from there?

https://github.com/dotnet/runtime/blob/210a7a9feec64b7fac5147656cddb199ec90cf75/src/coreclr/vm/assembly.cpp#L1424-L1437

noahfalk commented:
> So change this to something more generic and call both the RuntimeEventSource initialization and the startup hook initialization from there?

Yeah, that sounds good to me. Perhaps something like:

```C#
void ManagedStartup() // call this from native runtime startup code
{
    InitBeforeUserCode(); // run EventSource initialize in here
    ProcessStartupHooks(); // do the rest of the pre-existing startup hook logic here
}
```

lambdageek committed:
Revert "Make RuntimeEventSource.Initialize safe to call multiple times"

This reverts commit 07219db97e736bd49111dff8a605550909386b12.

lambdageek commented:
> > So change this to something more generic and call both the RuntimeEventSource initialization and the startup hook initialization from there?
> 
> Yeah, that sounds good to me. Perhaps something like:
> 
> ```cs
> void ManagedStartup() // call this from native runtime startup code
> {
>     InitBeforeUserCode(); // run EventSource initialize in here
>     ProcessStartupHooks(); // do the rest of the pre-existing startup hook logic here
> }
> ```

Hmm... that's actually not very linker friendly. Ideally I'd like to trim `ManagedStartup` if its body is entirely empty because both eventpipe and startup hooks are disabled.  Otherwise this could be a startup perf regression for Mono in published mobile/wasm apps.

lambdageek committed:
coreclr: Call RuntimeEventSource.Initialize from EventPipeAdapter::FinishInitialize

Decouple eventpipe initialization from startup hooks.

Previously if startup hooks were disabled, the runtime event source
would not be initialized.

lambdageek commented:
Okay, CoreCLR now follows Mono's model: `RuntimeEventSource.Initialize` is invoked from `EventPipeAdapter::FinishInitialization()` and is not related to startup hooks.

lambdageek commented on src/coreclr/vm/eventing/eventpipe/ep-rt-coreclr.cpp:
```diff
@@ -161,4 +162,23 @@ ep_rt_coreclr_sample_profiler_write_sampling_event_for_threads (
+
+        GCX_COOP();
+
+        // FIXME: how do I check if the method was trimmed away and avoid calling it?
+        MethodDescCallSite runtimeEventSourceInitialize(METHOD__RUNTIME_EVENT_SOURCE__INITIALIZE);
+        runtimeEventSourceInitialize.Call(NULL);
```
Question for CLR folks: I put `DEFINE_CLASS(RUNTIME_EVENT_SOURCE, ...)` wrapped in a `BEGIN_ILLINK_FEATURE_SWITCH` which means the method may be trimmed away if eventpipe is disabled.  How do I actually make this `MethodDescCallSite` lookup allowed to fail and avoid doing a call?

Or is this baking a method token right into the runtime?

vitek-karas replied:
In other places that I know of we don't rely on the lookup failing, instead we react to the feature switch in question and avoid doing the lookup completely. (For example built-in COM works this way)

lambdageek replied:
how can I check the value of the feature switch at runtime from native?

lambdageek replied:
I guess I can do `DEFINE_STATIC_PROPERTY` for `IsSupported` and call that and then call the Initialize method

lateralusX commented:
> @lateralusX @dotnet/dotnet-diag
> 
> So believe it or not the failure I'm seeing is due to the call in `StartupHookProvider` to `RuntimeEventSource.Initialize()`. Turns out mono already invokes that method from `ep_finish_init`. So when we run it a second time, we instantiate a second `RuntimeEventSource` instance and then things go wrong.
> 
> Questions:
> 
> 1. As far as I can tell, coreclr only calls `RuntimeEventSource.Initialize` from `StartupHookProvider`.  Which means if `!StartupHookProvider.IsSupported`, the runtime event source is not initialized? Is that intentional? I could make mono do the same thing, but it seems weird to couple these two features. Or is there a second call site in coreclr somewhere?
> 2. Alternately, I could make `RuntimeEventSource.Initialize` safe to call multiple times by only creating a new instance if there is not one already (probably using a lock-object variant of `LazyInitializer.EnsureInitialized`)
> 3. Alternately, I could add `#if !MONO` to StartupHookProvider and leave both runtimes with the same behavior we have today
> 
> ~**Update** I added option 2 to this PR, but we can go with another approach~ **Update 2** actually trying another option: Make `ep_rt_init_finish` for CoreCLR invoke `RuntimeEventSource.Initialize` same as in mono, and remove the init call from `StartupHookProvider`

On Mono we implemented init of `RuntimeEventSource` in Mono's shim layer implementation of `ep_rt_init_finish`:

https://github.com/dotnet/runtime/blob/d320fd90a2976b388169dd51b3732ad01bcb4266/src/mono/mono/eventpipe/ep-rt-mono.c#L2054

since we didn't have support for startup hooks. The implementation checks if the method have been trimmed out (feature been disabled) and won't do anything in that case.

noahfalk commented:
> Okay, CoreCLR now follows Mono's model: RuntimeEventSource.Initialize is invoked from EventPipeAdapter::FinishInitialization() and is not related to startup hooks.

Ideally RuntimeEventSource should be gated by EventSource.IsSupported, which isn't the same as whether EventPipe is supported. For example RuntimeEventSource could send some events via ETW or Lttng even if EventPipe had been removed from the runtime build.

If getting the code to be removed by the linker is at issue, can we hoist the checks to make it more obvious to the linker that code is removable? For example:

```C#
void ManagedStartup()
{
    if(EventSource.IsSupported) InitEventSource();
    if(StartupHookProvider.IsSupported) ProcessStartupHooks();
}
```

I'm fine with other ways of factoring it too as long as we can gate it with EventSource.IsSupported and not EventPipe.

lambdageek commented:
> If getting the code to be removed by the linker is at issue, can we hoist the checks to make it more obvious to the linker that code is removable? For example:

The issue is I want to avoid a call to `ManagedStartup` completely if both features are disabled.  But as far as I can tell, there's no way in a linker descriptor to gate a single method on a disjunction of two feature flags.

lambdageek committed:
Call RuntimeEventSource.Initialize only if EventSource.IsSupported

Also move the call out of EventPipeAdapter, back to ExecuteMainMethod.
EventSource is independent of EventPipe (for example if ETL is used,
or lttng)

lambdageek committed:
fix line damage

lambdageek committed:
more whitespace

lambdageek commented:
Well I ended up doing multiple calls from native to managed in CoreCLR.  But maybe this is a case where Mono and CoreCLR actually want different tradeoffs.  

I guess for CoreCLR the common scenario is no trimming or at least trimming with EventSource enabled and startup hooks disabled - so there's always going to be at least one call into managed no matter what.  So we might as well leave it in.  

On Mono, OTOH, we have wasm which will default to no event source and no startup hooks so the ideal case is 0 calls to managed.

So maybe I should add a managed `ManagedStartup` method for CoreCLR only, and leave mono with <=2 separate calls.

lambdageek committed:
Make a single managed startup method for CoreCLR

A common configuration for coreclr is event source enabled, startup
hooks disabled, so at least one managed call is inevitable.  Since we
have to call into managed no matter what, let the trimmer determine
what happens once we get there.

This is different from mono where published trimmed apps may have both
startup hooks and event source disabled.  In that case we would rather
avoid a managed call to an empty method early in startup.

lambdageek committed:
fix build and line damage

lambdageek committed:
fixup

noahfalk commented:
> Well I ended up doing multiple calls from native to managed in CoreCLR

That sounds fine to me, though the comment doesn't appear to match the current state of the PR? I'm fine if you want to leave the PR code as-is, or I'm also fine if you want to split into multiple calls from native->managed (presumably because that gives us a better optimization with features disabled, or better parity between Mono and CoreCLR).

lateralusX commented on src/mono/mono/metadata/object.c:
```diff
@@ -8124,6 +8124,25 @@ mono_runtime_get_managed_cmd_line (void)
+	if (mono_runtime_get_no_exec ())
+		return;
+
+	MonoClass *klass = mono_class_try_load_from_name (mono_defaults.corlib, "System", "StartupHookProvider");
```
Any reason why we won't call the same startup hook logic as CoreCLR and remove the call in ep_rt_mono_init_finish on Mono? If there might be a scenario in the future where the managed parts gets included without EventPipe on Mono (currenlty not the case), then we would need to change this since ep_finish_init would then just be a stub implementation.

lambdageek replied:
> Any reason why we won't call the same startup hook logic as CoreCLR and remove the call in ep_rt_mono_init_finish on Mono? If there might be a scenario in the future where the managed parts gets included without EventPipe on Mono (currenlty not the case), then we would need to change this since ep_finish_init would then just be a stub implementation., but the user still expect to get these events delivered over alternative provider. We could of course do that when needed, or we could just do it now, since we get the support that previously was missing and why init of runtime events ended up in ep_rt_mono_init_finish. Anything with the current startup hook implementation that would prevent us from getting the same behavior as we currently have, but done from startup hook instead of ep_rt_mono_init_finish?

There are two separate issues:

1. Support EventSource on Mono without EventPipe.
2. Use the same single `ManagedStartup` call as CoreCLR.

I think it's probably reasonable to do something about 1 - we can move the invoke of RuntimeEventSource.Initialize out of EP and into `mini_init`, for example.

For 2, the issue is that `ManagedStartup` cannot be trimmed - there is no way to tell the trimmer "remove a method if 2 feature switches are both disabled".  So we would always be paying for a call to managed even if event source and startup hooks are both disabled (for example on single-threaded WASM).

So if you think it makes sense to start decoupling EventSource and EventPipe in Mono now, we can do that.  But I don't think it makes sense to do 2.  Of course if we had some other managed initialization that had to be done explicitly on all platforms, then we could roll all that logic into one function.

lateralusX replied:
OK, understood, if we still won't end up on the same solution as CoreCLR, then I believe we can keep things as is for now, and move the call to init when/if we ever have a need to separate init of runtime event source from EventPipe.

lateralusX reviewed (approved):
Mono bits LGTM!

lambdageek commented:
Verified manually (by overwriting files in runtime packs) that the Microsoft.Android SDK works with this implementation of startup hooks
<img width="989" alt="Screenshot 2023-01-18 at 12 13 01" src="https://user-images.githubusercontent.com/480437/213251107-d9413f47-7a40-4a6c-8ae1-6a3d992b5820.png">

lambdageek commented:
linux-arm64 llvmfullaot timeout is known
